source/server/filter_chain_manager_impl.cc

```diff
@@ -496,6 +571,19 @@ FilterChainManagerImpl::findFilterChain(const Network::ConnectionSocket& socket)
              : default_filter_chain_.get();
 }
 
+const Network::FilterChain*
+FilterChainManagerImpl::findFilterChainUsingMatcher(const Network::ConnectionSocket& socket) const {
```

Comment:
nit: Currently a failed match is indistinguishable from the action FilterChain not being found (with default_filter_chain unset), even at debug log level. It may be helpful to have some sort of logging to help distinguish these, which is often user request error (sending wrong request and failing to match) vs config/control plane error (failed to send correct Filter Chain to go with the match)

Reply:
Scope creeping this a ton, for development purposes it could be nice to have some sort of debugging info on where in the tree we got. For some especially deep matching trees getting a binary "no match" can be pretty hard to decipher. It would be really cool if there was some sort of log like "match port,protocol,ip but failed to match on sourcePort" (where those things are the matcher names).

Definitely not a blocker though

Reply:
Does the listener tcp access log with filter chain name not show the chosen filter chain? Extra trace logging in the matcher framework would also be fine.

Reply:
That works for matching, I was thinking more about when it doesn't match. I suppose both are useful but would guess debugging non-match is more common.

I was playing with this PR earlier and got stuck on two issues (with my config, not the code here). Both ended up being pointing to filter chains that do not exist; so really the distinguishing of "no match" and "invalid action" would have been enough to help me debug the issues I have faced so far

Reply:
I see. Let me think about it. This was intended for an eventually consistent model where some filter chains might arrive later. Right now we process everything in a batch so can easily log stuff.